sparse/src/KokkosSparse_CcsMatrix.hpp

```diff
@@ -0,0 +1,265 @@
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+/// \file Kokkos_Sparse_CcsMatrix.hpp
+/// \brief Local sparse matrix interface
+///
+/// This file provides KokkosSparse::CcsMatrix.  This implements a
+/// local (no MPI) sparse matrix stored in compressed row sparse
+/// ("Ccs") format.
+
+#ifndef KOKKOS_SPARSE_CCSMATRIX_HPP_
+#define KOKKOS_SPARSE_CCSMATRIX_HPP_
+
+#include "Kokkos_Core.hpp"
+#include "Kokkos_StaticCrsGraph.hpp"
+#include <sstream>
+#include <stdexcept>
+#include <type_traits>
+#include "KokkosSparse_findRelOffset.hpp"
+#include "KokkosKernels_default_types.hpp"
+#include "KokkosKernels_Macros.hpp"
+
+namespace Kokkos {
+/// \class StaticCcsGraph
+/// \brief Compressed column storage array copied from Kokkos::StaticCrsGraph.
+///
+/// \tparam DataType The type of stored entries.  If a StaticCrsGraph is
+///   used as the graph of a sparse matrix, then this is usually an
+///   integer type, the type of the column indices in the sparse
+///   matrix.
+///
+/// \tparam Arg1Type The second template parameter, corresponding
+///   either to the Device type (if there are no more template
+///   parameters) or to the Layout type (if there is at least one more
+///   template parameter).
+///
+/// \tparam Arg2Type The third template parameter, which if provided
+///   corresponds to the Device type.
+///
+/// \tparam Arg3Type The third template parameter, which if provided
+///   corresponds to the MemoryTraits.
+///
+/// \tparam SizeType The type of col offsets.  Usually the default
+///   parameter suffices.  However, setting a nondefault value is
+///   necessary in some cases, for example, if you want to have a
+///   sparse matrices with dimensions (and therefore column indices)
+///   that fit in \c int, but want to store more than <tt>INT_MAX</tt>
+///   entries in the sparse matrix.
+///
+/// A col has a range of entries:
+/// <ul>
+/// <li> <tt> col_map[i0] <= entry < col_map[i0+1] </tt> </li>
+/// <li> <tt> 0 <= i1 < col_map[i0+1] - col_map[i0] </tt> </li>
+/// <li> <tt> entries( entry ,            i2 , i3 , ... ); </tt> </li>
+/// <li> <tt> entries( col_map[i0] + i1 , i2 , i3 , ... ); </tt> </li>
+/// </ul>
+template <class DataType, class Arg1Type, class Arg2Type = void,
+          class Arg3Type    = void,
+          typename SizeType = typename ViewTraits<DataType*, Arg1Type, Arg2Type,
+                                                  Arg3Type>::size_type>
+class StaticCcsGraph
```

Comment:
@e10harvey If they are going to be part of the same class hierarchy, CrsMatrix and CcsMatrix should definitely be siblings and not parent and child. It's more intuitive to have them on equal footing, and if someone writes
```
void foo(const CrsMatrix& A) {...}
```
it should not be possible to accidentally pass an instance of CcsMatrix for A. Having them be siblings lets you handle Crs/Ccs with overloading or by taking CompressedSparseMatrix and dynamic casting.

Reply:
@brian-kelley: Excellent point. Along those same lines, rather than copying and pasting StaticCrsGraph from Kokkos I'd like to keep what is currently here but move StaticCcsGraph into an impl namespace:
```
namespace KokkosKernels {
namespace Impl {
template ...
class StaticCcsGraph : StaticCrsGraph
...
}
}
```
Any objections?

Reply:
I'm not sure how you would rename ``row_map`` to ``col_map`` if you do it this way.

Reply:
Aliasing col_map to the parent's row_map is the best I could come up with:
```c++
class StaticCcsGraph
    : public Kokkos::StaticCrsGraph<DataType, Arg1Type, Arg2Type, Arg3Type,
                                    SizeType> {
 private:
  using parent_type =
      Kokkos::StaticCrsGraph<DataType, Arg1Type, Arg2Type, Arg3Type, SizeType>;

 public:
  using parent_type::parent_type;
  using col_map_type = typename parent_type::row_map_type;

  col_map_type& col_map = parent_type::row_map;

  KOKKOS_INLINE_FUNCTION
  typename parent_type::size_type numCols() const {
    return parent_type::numRows();
  }
};
```

I would prefer to create a row / col agnostic StaticCsGraph and inherit from it but Kokkos has no use-case for this, just StaticCrsGraph.

Reply:
Ah, ok. With this I'm now thinking that it will still let you call ``numRows()``, but that shouldn't exist as a function. But then if you did protected/private inheritance, it would hide ``entries`` which we do want to expose.

Reply:
OK, that's actually not a problem, you can do ``using parent_type::entries`` under ``public:``

Reply:
I ended up adding a bare-bones StaticCcsGraph instead of using inheritance. Sorry for all the confusion and back-and-forth on what should have been a simple feature request. I guess I'm still coming up-to-speed.